<?xml version="1.0"?>
<robot name="mecanum_drive_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- Dimensions du corps -->
    <xacro:property name="body_link_x_dim" value="1.0"/>
    <xacro:property name="body_link_y_dim" value="0.6"/>
    <xacro:property name="body_link_z_dim" value="0.3"/>

    <!-- Dimensions des roues -->
    <xacro:property name="wheel_radius" value="0.1"/>
    <xacro:property name="wheel_width" value="0.05"/>

    <!-- Densité du matériau -->
    <xacro:property name="body_density" value="2710.0"/>
    <xacro:property name="wheel_density" value="2710.0"/>

    <!-- Masse du robot et des roues -->
    <xacro:property name="body_mass" value="${body_density * body_link_x_dim * body_link_y_dim * body_link_z_dim}"/>
    <xacro:property name="wheel_mass" value="${wheel_density * 3.14159265 * wheel_radius * wheel_radius * wheel_width}"/>

    <!-- Inertie des roues -->
    <xacro:property name="I_wheel" value="${(1/12) * wheel_mass * (3 * wheel_radius * wheel_radius + wheel_width * wheel_width)}"/>
    <xacro:property name="Iz_wheel" value="${0.5 * wheel_mass * wheel_radius * wheel_radius}"/>

    <!-- Définir les liens et les joints -->
    <link name="body_link">
        <visual>
            <geometry>
                <box size="${body_link_x_dim} ${body_link_y_dim} ${body_link_z_dim}"/>
            </geometry>
            <origin xyz="0 0 ${body_link_z_dim/2}" rpy="0 0 0"/>
        </visual>
        <collision>
            <geometry>
                <box size="${body_link_x_dim} ${body_link_y_dim} ${body_link_z_dim}"/>
            </geometry>
            <origin xyz="0 0 ${body_link_z_dim/2}" rpy="0 0 0"/>
        </collision>
        <inertial>
            <origin xyz="0 0 ${body_link_z_dim/2}" rpy="0 0 0"/>
            <mass value="${body_mass}"/>
            <inertia
                ixx="${(1/12)*body_mass*(body_link_y_dim*body_link_y_dim+body_link_z_dim*body_link_z_dim)}"
                ixy="0"
                ixz="0"
                iyy="${(1/12)*body_mass*(body_link_x_dim*body_link_x_dim+body_link_z_dim*body_link_z_dim)}"
                iyz="0"
                izz="${(1/12)*body_mass*(body_link_y_dim*body_link_y_dim+body_link_x_dim*body_link_x_dim)}"/>
        </inertial>
    </link>

    <!-- Définition d'une macro pour l'inertie des roues -->
    <xacro:macro name="wheel_inertial">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${wheel_mass}"/>
            <inertia
                ixx="${I_wheel}"
                ixy="0.0"
                ixz="0.0"
                iyy="${I_wheel}"
                iyz="0.0"
                izz="${Iz_wheel}"/>
        </inertial>
    </xacro:macro>

    <!-- Définition des roues et des joints -->
    <link name="wheel1_link">
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </collision>
        <xacro:wheel_inertial/>
    </link>

    <link name="wheel2_link">
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </collision>
        <xacro:wheel_inertial/>
    </link>

    <link name="wheel3_link">
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </collision>
        <xacro:wheel_inertial/>
    </link>

    <link name="wheel4_link">
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 1.5708 0"/>
        </collision>
        <xacro:wheel_inertial/>
    </link>

    <!-- Joints des roues -->
    <joint name="wheel1_joint" type="continuous">
        <parent link="body_link"/>
        <child link="wheel1_link"/>
        <origin xyz="${body_link_x_dim/2 - wheel_radius} ${body_link_y_dim/2 + wheel_width/2} ${wheel_radius}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="1000" velocity="1000"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <joint name="wheel2_joint" type="continuous">
        <parent link="body_link"/>
        <child link="wheel2_link"/>
        <origin xyz="${body_link_x_dim/2 - wheel_radius} ${-body_link_y_dim/2 - wheel_width/2} ${wheel_radius}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="1000" velocity="1000"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <joint name="wheel3_joint" type="continuous">
        <parent link="body_link"/>
        <child link="wheel3_link"/>
        <origin xyz="${-body_link_x_dim/2 + wheel_radius} ${body_link_y_dim/2 + wheel_width/2} ${wheel_radius}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="1000" velocity="1000"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <joint name="wheel4_joint" type="continuous">
        <parent link="body_link"/>
        <child link="wheel4_link"/>
        <origin xyz="${-body_link_x_dim/2 + wheel_radius} ${-body_link_y_dim/2 - wheel_width/2} ${wheel_radius}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="1000" velocity="1000"/>
        <dynamics damping="1.0" friction="1.0"/>
    </joint>

    <!-- Matériaux et friction -->
    <gazebo reference="wheel1_link">
        <material>Gazebo/Blue</material>
        <kp>1000000.0</kp>
        <kd>100.0</kd>
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
    </gazebo>

    <gazebo reference="wheel2_link">
        <material>Gazebo/Green</material>
        <kp>1000000.0</kp>
        <kd>100.0</kd>
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
    </gazebo>

    <gazebo reference="wheel3_link">
        <material>Gazebo/Yellow</material>
        <kp>1000000.0</kp>
        <kd>100.0</kd>
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
    </gazebo>

    <gazebo reference="wheel4_link">
        <material>Gazebo/Red</material>
        <kp>1000000.0</kp>
        <kd>100.0</kd>
        <mu1>0.9</mu1>
        <mu2>0.9</mu2>
    </gazebo>

    <!-- Include the file that defines extra Gazebo options and motion control driver -->
    <xacro:include filename="$(find robot_model_one_pkg)/urdf/robot.gazebo" />
</robot>

